packages/SwingSet/docs/async.md

```diff
@@ -0,0 +1,91 @@
+# Asynchronous Considerations in SwingSet
+
+When working with SwingSet, it's crucial to understand how asynchronous operations interact with the system's execution model. A **vat** (a unit of isolated execution in SwingSet) may restart or terminate before an asynchronous operation completes. Although JavaScript's `await` keyword allows you to write asynchronous code that resembles synchronous code, you must be mindful of events that could prevent a promise from settling:
+
+- **Underlying condition never satisfied**: The condition required for the promise to resolve may never occur.
+- **Vat restarts or upgrades before settlement**: The promise may be severed if the vat restarts or upgrades before it settles.
+- **Vat terminates**: If the vat terminates, execution ends entirely, and the promise will not settle.
+
+# Promise Temporality
+
+When programming with promises in SwingSet, it's important to understand the **temporality** of promise settlement—that is, the timing and execution dependencies required for a promise to settle. Promises can be classified into three categories based on their temporality: **Immediate**, **Prompt**, and **Delayed**.
+
+## Immediate Promises
+
+An **Immediate** promise is one where all the necessary state for it to settle is available at the time of its creation. As a result, it will settle within the same event loop iteration (or **"crank"** in SwingSet terminology) in which it was created. We refer to this as "Immediate" because it occurs before any asynchronous callbacks, similar to how operations happen before `setImmediate` callbacks in JavaScript.
+
+### Characteristics of Immediate Promises
+
+- **Settlement Timing**: Settles within the same crank it was created.
+- **Dependencies**: Requires no additional state or input; all information is already available.
+- **Safety**: Not affected by vat restarts or upgrades within the same crank.
+
+## Prompt Promises
+
+A **Prompt** promise settles without requiring further input from outside SwingSet. All Immediate promises are also Prompt, but not all Prompt promises are Immediate.
+
+The SwingSet kernel processes events until there is nothing left to run, then handles the next I/O event and repeats the loop. A promise is considered Prompt if it resolves before new I/O occurs, and thus before any vat upgrades (since only I/O triggers vat upgrades).
```

Comment:
I think other swingset docs aren't phrased in terms of I/O but rather turning control over to the host.
cc @warner